File: Data Structures & Algorithms/house-robber/submission-3.cpp
class Solution {
public:
vector<int> dp;
    int rob(vector<int>& nums) {
        dp.resize(nums.size(),-1);
        return dfs(nums,0);
    }
    int dfs(vector<int>& nums,int i){
       if(i>=nums.size())return 0;
       if(dp[i]!=-1)return dp[i];

       dp[i]= max(dfs(nums,i+1),nums[i]+dfs(nums,i+2));
       return dp[i];
       
    }
};

File: Data Structures & Algorithms/target-sum/submission-3.java
class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        int sum=0;
        for(int i=0;i<nums.length;i++){
            sum+=nums[i];
        }
        int dp[][] = new int[nums.length+1][2*sum+1];
        for(int i=0;i<=nums.length;i++){
            for(int j=0;j<2*sum+1;j++){
                dp[i][j]=-1;
            }
        }
        return dfs(nums,target,0,dp,sum);
    }
    int dfs(int[] nums , int target,int index,int[][] dp,int sum){
           if(index==nums.length && target==0){
            
            return 1;
           }
           if(index==nums.length && target!=0){
            return 0;
           }
           int col = target+sum;
           if (target < -sum || target > sum) return 0;
           if(dp[index][col]!=-1)return dp[index][col];
           int plus = dfs(nums,target-nums[index],index+1,dp,sum);
           int sub = dfs(nums,target+nums[index],index+1,dp,sum);
           return dp[index][col] =plus+sub;
    }
}
